### PR TITLE
@wordpress/data: Handle more return type values from a mapSelect argument on useSelect

### LINKED ISSUE #16612: useSelect doesn't return the updated value if it is not an object.
**Describe the bug**
Following the documentation of [useSelect hook](https://github.com/WordPress/gutenberg/tree/master/packages/data#useSelect) to return a value from the store (custom or core) the updated value is only returned if the callback returns an object.

**To reproduce**
Try the following **isolated**. Notice that in the third example `getSelectedBlock()` is being returned as is. 

Doesn't update
`const is_typing = useSelect(select => select("core/block-editor").isTyping());`

Updates correctly
`const is_typing = useSelect(select => ({ value: select("core/block-editor").isTyping() }));`

Updates correctly
`const selected = useSelect(select => select("core/block-editor").getSelectedBlock());`

**Version:**
 - WordPress: 5.2.2
 - Gutenberg 6.1.1

### PR DESCRIPTION
**Pull request by nerrad**
## Description

See #16612 for original report.  The `useSelect` hook uses `isShallowEqualObjects` for comparing whether the new value returned from the `mapSelect` callback is equal to the previous value. See 

https://github.com/WordPress/gutenberg/blob/a894b3520598eceb91ddd8bb0299024b4f970afa/packages/data/src/components/use-select/index.js#L132-L134

I added some unit tests that demonstrated (when run against original `useSelect`) if a `mapSelect` callback returns a number or boolean primitive type, it will not get handled as expected by the `useSelect` hook.  Oddly enough, `array` and `string` types did get handled fine.  The same tests pass with the work done in this branch.

This pull adds a new utility function `isEqual` (which is not exposed on a public api so should be okay to use naming wise) for handling all primitive types and doing the appropriate comparison.  This is then implemented in `useSelect`.

Fixes #16612.

## How has this been tested?

* [x] New unit tests add coverage for this, and existing unit tests pass just fine.
* [x] Any serious problems should be picked up by e2e tests as well.

## Types of changes

This is a non-breaking change that simply fixes a problem with expectations (especially considering the current documentation) around what a `mapSelect` callback given as an argument on `useSelect` can return.  It can now return a primitive value, an array, or an object.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by nerrad** [comment]
@epiqueras would love it if you could review this!

File: packages/data/src/components/use-select/test/index.js [hunk 1]
```diff
@@ -132,5 +132,89 @@ describe( 'useSelect', () => {
 			children: 'bar',
 		} );
 	} );
+	describe( 'rerenders as expected with various mapSelect return ' +
+		'types', () => {
+		const getComponent = ( mapSelectSpy, dependencyKey ) => ( props ) => {
```
Comment (epiqueras) [inline packages/data/src/components/use-select/test/index.js]: These don't really need to be mocks and using normal functions would make the test easier to read.

We should only use mocks when we need to read from the `.mock` property. E.g. to introspect on what parameters the function was called with.

File: packages/data/src/components/use-select/test/index.js [hunk 2]
```diff
@@ -132,5 +132,89 @@ describe( 'useSelect', () => {
...
+		} );
+		[
+			[
+				'boolean',
+				[ false, true ],
+			],
+			[
+				'number',
+				[ 10, 20 ],
+			],
```
Comment (epiqueras) [inline packages/data/src/components/use-select/test/index.js]: We should also test objects and null <=> undefined.

File: packages/data/src/components/use-select/test/index.js (hunk 1 above)
Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: I converted two of the mocks that were unnecessary in 511b504.  However, the others still had utility for both spying on the calls and changing the return value.

File: packages/data/src/components/use-select/test/index.js (hunk 2 above)
Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: Added additional tests in de7f4fc (modification to one in 46f80f6) for covering null/undefined/objects.  Had to change the test component so tests would validate on prop rather than children (because objects cannot be returned as the value for children).

File: packages/data/src/components/use-select/test/index.js [hunk 3]
```diff
@@ -132,5 +132,93 @@ describe( 'useSelect', () => {
 			children: 'bar',
 		} );
 	} );
+	describe( 'rerenders as expected with various mapSelect return ' +
+		'types', () => {
+		const getComponent = ( mapSelectSpy, dependencyKey ) => ( props ) => {
+			const dependencies = props[ dependencyKey ];
+			mapSelectSpy.mockImplementation(
```
Comment (epiqueras) [inline packages/data/src/components/use-select/test/index.js]: This implementation never needs to change. It can be a regular function.

File: packages/data/src/components/use-select/test/index.js [hunk 4]
```diff
@@ -132,5 +132,93 @@ describe( 'useSelect', () => {
...
+				( select ) => select( 'testStore' ).testSelector( props.keyName )
```
Comment (epiqueras) [inline packages/data/src/components/use-select/test/index.js]: The parameter in the selector is unused.

File: packages/data/src/components/use-select/test/index.js [hunk 5]
```diff
@@ -132,5 +132,93 @@ describe( 'useSelect', () => {
...
+							<RegistryProvider value={ registry }>
+								<TestComponent keyName="foo" />
+							</RegistryProvider>
+						);
+					} );
+					const testInstance = renderer.root;
+					// ensure expected state was rendered.
+					expect( testInstance.findByType( 'div' ).props.data ).toEqual( valueA );
+
+					// update the returned valud from the selector and trigger the
```
Comment (epiqueras) [inline packages/data/src/components/use-select/test/index.js]:
```suggestion
					// Update the returned value from the selector and trigger the
```

File: packages/data/src/components/use-select/test/index.js (hunk 3 above)
Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: Do you mean `mapSelectSpy` can be a regular function?  The problem with making this change is that the test _is_ spying on the implementation here:https://github.com/WordPress/gutenberg/pull/16669/files#diff-bfc6925e9604200fafd2b38e12993a97R218

Comment (epiqueras) [reply packages/data/src/components/use-select/test/index.js]: Oh, I see, maybe it should have the same variable name to avoid confusion. They are sibling scopes so it wouldn't be shadowing.

File: packages/data/src/components/use-select/test/index.js (hunk 4 above)
Comment (epiqueras) [reply packages/data/src/components/use-select/test/index.js]: We can remove all the usages of `dependencyKey` and `keyName` here. The test never changes the dependency so it would be just like passing an empty array instead.

File: packages/data/src/components/use-select/test/index.js (hunk 3 above)
Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: See f3fe2e0

File: packages/data/src/components/use-select/test/index.js (hunk 4 above)
Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: see f3fe2e0

**Comment by nerrad** [comment]
I don't think I've had this much back and forth on _tests_ before 😃 !  All good though.

**Comment by epiqueras** [comment]
Tests are not only for validating logic and guarding regressions. They serve as living documentation of our code for folks who are adventuring into parts of the codebase that are unknown to them. I don't think they should have different quality standards than our other code.

That said, the test is looking a lot better after the initial iterations. There are still a few things I would change like the unnecessary `.mockImplementation` call that could be replaced by an inline definition in the spy's declaration and the confusing naming of some functions that aren't actually spies as spies. But I wouldn't say that these are blockers and the code is much easier to read now.

Thanks for taking care of this PR :smile:

**Comment by nerrad** [comment]
> Tests are not only for validating logic and guarding regressions. They serve as living documentation of our code for folks who are adventuring into parts of the codebase that are unknown to them. I don't think they should have different quality standards than our other code.

I 100% agree.  My comment wasn't a complaint.  It was more of an observation of the thoroughness of your review (hence the smiley).  I probably could've worded better though.

File: packages/data/src/components/use-select/test/index.js [hunk 6]
```diff
@@ -132,5 +132,91 @@ describe( 'useSelect', () => {
...
+			],
+			[
+				'null',
+				[ null, undefined ],
+			],
+			[
+				'undefined',
+				[ undefined, 42 ],
+			],
+		].forEach(
```
Comment (swissspidy) [inline packages/data/src/components/use-select/test/index.js]: Why not use the `each([...]).test` approach that Jest supports natively?

**Comment by epiqueras** [comment]
>I 100% agree. My comment wasn't a complaint. It was more of an observation of the thoroughness of your review (hence the smiley). I probably could've worded better though.

I didn't assume it was, I try to not make assumptions like that, because so much tone and context is lost through text. But I figured it wouldn't hurt to make my approach as transparent as possible :smile:

File: packages/data/src/components/use-select/test/index.js (hunk 6 above)
Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: Honestly, I just didn't think to use it :).  It does look like the test could be switched to use it here but is it functionally any different?  Are there some advantages to using it here?

Comment (swissspidy) [reply packages/data/src/components/use-select/test/index.js]: Different reporting I think, and better ESLint support.

Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: Interesting, good points and I'll give a go at updating the tests to use the built-in test conditions api in jest.  It's also something I'm willing to do in a separate pull if this fix is needed to move #16460 along.

File: packages/data/src/components/use-select/utils.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/**
+ * WordPress dependencies
+ */
+import isShallowEqual from '@wordpress/is-shallow-equal';
+
+const isEqual = ( valueA, valueB ) => {
```
Comment (aduth) [inline packages/data/src/components/use-select/utils.js]: Might be worth a brief JSDoc to explain the purpose we're seeking to accomplish with this function.

Comment (ellatrix) [reply packages/data/src/components/use-select/utils.js]: When I see `isEqual`, I think of a deep comparison. Might be good to specify what kind of equality this checks for.

Comment (nerrad) [reply packages/data/src/components/use-select/utils.js]: Agreed to both points, I'll try and get this added today sometime.

Comment (epiqueras) [reply packages/data/src/components/use-select/utils.js]: We should just use the default export of the shallow equality package.

```
import isShallowEqual from '@wordpress/is-shallow-equal';
```

I am pretty sure it does the same thing.

Comment (nerrad) [reply packages/data/src/components/use-select/utils.js]: doh ya, tests pass with just `isShallowEqual`... I'll update and great catch!

Comment (nerrad) [reply packages/data/src/components/use-select/utils.js]: > We should just use the default export of the shallow equality package.

Implemented in 642d672

File: packages/data/src/components/use-select/test/index.js (hunk 6 above)
Comment (nerrad) [reply packages/data/src/components/use-select/test/index.js]: implemented in e2f6c28

File: packages/data/src/components/use-select/utils.js (hunk 7 above)
Comment (aduth) [reply packages/data/src/components/use-select/utils.js]: > We should just use the default export of the shallow equality package.
> 
> ```
> import isShallowEqual from '@wordpress/is-shallow-equal';
> ```
> 
> I am pretty sure it does the same thing.

It's true and I think it's fine to ship with it as-is (especially with tests to guard against regressions), and while technically it's the specified behavior, it's probably not exceedingly obvious that in code here we're intentionally anticipating to handle non-object/array values. It could be worth a clarifying comment to that end.

---

_Aside:_ `isShallowEqual` is documented as requiring the arguments be passed:

https://github.com/WordPress/gutenberg/blob/fcbdd20ee17b298ac236f71dc4102020882216df/packages/is-shallow-equal/index.js#L15-L16

But it's tested to handle others:

https://github.com/WordPress/gutenberg/blob/fcbdd20ee17b298ac236f71dc4102020882216df/packages/is-shallow-equal/test/index.js#L136-L147

It should probably allow any type.

**Review by aduth — APPROVED** [review]
